src/mlpack/core/cv/metrics/f1.hpp

```diff
@@ -62,6 +62,43 @@ class F1
    * to maximize the metric.
    */
   static const bool NeedsMinimization = false;
+
+ private:
+  /**
+   * Run classification and calculate F1 for binary classification.
+   */
+  template<AverageStrategy _AS,
+           typename MLAlgorithm,
+           typename DataType,
+           typename = std::enable_if_t<_AS == Binary>>
+  static double Evaluate(MLAlgorithm& model,
+                        const DataType& data,
+                        const arma::Row<size_t>& labels);
+
+  /**
+   * Run classification and calculate microaveraged F1.
+   */
+  template<AverageStrategy _AS,
+           typename MLAlgorithm,
+           typename DataType,
+           typename = std::enable_if_t<_AS == Micro>,
+           typename = void>
+  static double Evaluate(MLAlgorithm& model,
+                        const DataType& data,
+                        const arma::Row<size_t>& labels);
+
+  /**
+   * Run classification and calculate macroaveraged F1.
+   */
+  template<AverageStrategy _AS,
+           typename MLAlgorithm,
+           typename DataType,
+           typename = std::enable_if_t<_AS == Macro>,
+           typename = void,
+           typename = void>
+  static double Evaluate(MLAlgorithm& model,
+                        const DataType& data,
+                        const arma::Row<size_t>& labels);
```

Comment:
Wow, the refactoring necessary to allow the user to specify a positive class is a lot harder than I thought!  I think that you might be able to use SFINAE in the arguments instead of adding the extra `typename = void`s:

```
template<AverageStrategy _AS, typename MLAlgorithm, typename DataType>
static double Evaluate(
    MLAlgorithm& model,
    const DataType& data,
    const arma::Row<size_t>& labels,
    const std::enable_if_t<_AS == Micro>* = 0);
```

and likewise for `_AS == Macro` and `_AS == Binary`.